codex/stores/maintenance.nim

```diff
@@ -10,92 +10,318 @@
+proc decode(T: type DatasetMetadata, bytes: seq[byte]): ?!T = T.fromJson(bytes)
+
+proc trackExpiry*(
+  self: DatasetMaintainer,
+  treeCid: Cid,
+  leavesCount: Natural,
+  expiry: SecondsSince1970,
+  manifestsCids: seq[Cid] = @[]
+): Future[?!void] {.async.} =
+  # Starts tracking expiry of a given dataset
+  #
+
+  trace "Tracking an expiry of a dataset", treeCid, leavesCount, expiry
+
+  without key =? createDatasetMetadataKey(treeCid), err:
+    return failure(err)
+
+  await modify[DatasetMetadata](self.metaDs, key,
+    proc (maybeCurrDatasetMd: ?DatasetMetadata): Future[?DatasetMetadata] {.async.} =
+      var md: DatasetMetadata
+
+      if currDatasetMd =? maybeCurrDatasetMd:
+        md.expiry = max(currDatasetMd.expiry, expiry)
+
+        if currDatasetMd.leavesCount != leavesCount:
+          raise newException(CatchableError, "DatasetMetadata for treeCid " & $treeCid & " is already stored with leavesCount " &
+            $currDatasetMd.leavesCount & ", cannot override it with leavesCount " & $leavesCount)
+
+        md.leavesCount = leavesCount
+        md.manifestsCids = (currDatasetMd.manifestsCids & manifestsCids).deduplicate
+        md.checkpoint = Checkpoint(progress: 0, timestamp: 0)
+      else:
+        md.expiry = expiry
+        md.leavesCount = leavesCount
+        md.manifestsCids = manifestsCids
+        md.checkpoint = Checkpoint(progress: 0, timestamp: 0)
+
+      md.some
+  )
+
+proc trackExpiry*(
+  self: DatasetMaintainer,
+  treeCid: Cid,
+  leavesCount: Natural,
+  manifestsCids: seq[Cid] = @[]
+): Future[?!void] {.async.} =
+  await self.trackExpiry(treeCid, leavesCount, self.clock.now + self.defaultExpiry.seconds, manifestsCids)
+
+proc trackExpiry*(
+  self: DatasetMaintainer,
+  cid: Cid,
+  manifestsCids: seq[Cid] = @[]
+): Future[?!void] {.async.} =
+  await self.trackExpiry(cid, 0, self.clock.now + self.defaultExpiry.seconds, manifestsCids)
+
+proc ensureExpiry*(
+  self: DatasetMaintainer,
+  treeCid: Cid,
+  minExpiry: SecondsSince1970): Future[?!void] {.async.} =
+  ## Sets the dataset expiry to a max of two values: current expiry and `minExpiry`,
+  ## if a dataset for given `treeCid` is not currently tracked a CatchableError is thrown
+  ##
+
+  trace "Updating a dataset expiry", treeCid, minExpiry
+
+  without key =? createDatasetMetadataKey(treeCid), err:
+    return failure(err)
+
+  await modify[DatasetMetadata](self.metaDs, key,
+    proc (maybeCurrDatasetMd: ?DatasetMetadata): Future[?DatasetMetadata] {.async.} =
+      if currDatasetMd =? maybeCurrDatasetMd:
+        let datasetMd = DatasetMetadata(
+          expiry: max(currDatasetMd.expiry, minExpiry),
+          leavesCount: currDatasetMd.leavesCount,
+          manifestsCids: currDatasetMd.manifestsCids,
+          checkpoint: currDatasetMd.checkpoint
+        )
+        return datasetMd.some
+      else:
+        raise newException(CatchableError, "DatasetMetadata for treeCid " & $treeCid & " not found")
+  )
```

Comment:
> If we want a modify operation to be stopped, raising an exception is the only
> way to do that

With Dmitriy's suggested change, returning a Result will be the way to stop an
operation, which is exactly what I had in mind.

> I can extract that anonymous proc into a named proc if you want, however I
> don't see how it automatically becomes more readable this way.

Understood. As an outside reader, I thought perhaps you might want to know what
is considered subjectively "more readable" for that reader.

Reply:
>You say this because of the contract for [modifyGet](https://github.com/codex-storage/nim-datastore/blob/f4989fcce5d74a648e7e2598a72a7b21948f4a85/datastore/datastore.nim#L57), right? Cause there is currently no provisioning there for a modify operation to be aborted?

It's not mention in the docs, but in the signature, we return Future which implies it can be a failure.

>I suppose returning maybeCurrDatasetMd would be equivalent to a NOP, but sort of inefficient as it would still trigger a write to the underlying store?

Yep, returning the original argument is NOP. So the state in datastore is essentially the same as raising error. The difference is that when error is raised from a closure, that error is propagated to the caller of `modifyGet` as `failure(err)`. Probably it should be documented.

Reply:
@emizzle 
>With Dmitriy's suggested change, returning a Result will be the way to stop an
operation, which is exactly what I had in mind.

Future already communicates the error. With `Future[Result[T]]` it becomes ambiguous where the error will be. 

Also not sure why are we even talking about it. I'm not changing anything there.